### PR TITLE
[CALCITE] Added support for UPPERCASE and CASESPECIFIC column level attributes for CREATE TABLE expression

### PR DESCRIPTION
**Pull request by AndrewPochapsky**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/SqlDataTypeSpec.java [hunk 1]
```diff
@@ -153,13 +167,42 @@ public Boolean getNullable() {
...
+  public SqlDataTypeSpec withUppercase(Boolean uppercase) {
+    if (Objects.equals(uppercase, this.uppercase)) {
+      return this;
+    }
+    return new SqlDataTypeSpec(typeNameSpec, timeZone, nullable, uppercase,
+        caseSpecific, getParserPosition());
+  }
+
+  /** Returns a copy of this data type specification with a given
+   * nullability. */
```
Comment (michaelvdow) [inline core/src/main/java/org/apache/calcite/sql/SqlDataTypeSpec.java:199]: I think this comment could be updated for case specific instead of nullability

Comment (AndrewPochapsky) [reply core/src/main/java/org/apache/calcite/sql/SqlDataTypeSpec.java:199]: You're correct, darn copy and paste!

File: babel/src/main/codegen/includes/parserImpls.ftl [hunk 2]
```diff
@@ -133,23 +133,83 @@ SqlNodeList ExtendColumnList() :
     }
 }
 
+boolean ProcessNull() :
```
Comment (dasch-1) [inline babel/src/main/codegen/includes/parserImpls.ftl:136]: /ProcessNull/IsNullable/

File: babel/src/main/codegen/includes/parserImpls.ftl [hunk 3]
```diff
@@ -133,23 +133,83 @@ SqlNodeList ExtendColumnList() :
...
+            return false;
+        }
+    |
+        <NULL> {
+            return true;
+        }
+    )
+}
+
+boolean ProcessCaseSpecific() :
```
Comment (dasch-1) [inline babel/src/main/codegen/includes/parserImpls.ftl:152]: /ProcessCaseSpecific/IsCaseSpecific/

File: babel/src/main/codegen/includes/parserImpls.ftl [hunk 4]
```diff
@@ -133,23 +133,83 @@ SqlNodeList ExtendColumnList() :
...
+            return false;
+        }
+    |
+        <CASESPECIFIC> {
+            return true;
+        }
+    )
+}
+
+boolean ProcessUpperCase() :
```
Comment (dasch-1) [inline babel/src/main/codegen/includes/parserImpls.ftl:168]: /ProcessUpperCase/IsUpperCase/

File: babel/src/main/codegen/includes/parserImpls.ftl [hunk 5]
```diff
@@ -133,23 +133,83 @@ SqlNodeList ExtendColumnList() :
...
+    Boolean caseSpecific = null;
     final Span s = Span.of();
 }
 {
     id = CompoundIdentifier()
     type = DataType()
-    [
-        <NOT> <NULL> {
-            nullable = false;
+    //This acts as a loop to check which optional parameters have been specified
```
Comment (dasch-1) [inline babel/src/main/codegen/includes/parserImpls.ftl:196]: nit: Missing a space between slash and first word.

Comment (dasch-1) [inline babel/src/main/codegen/includes/parserImpls.ftl:196]: nit: Suggest ending comments with punctuation.

File: babel/src/test/java/org/apache/calcite/test/BabelParserTest.java [hunk 6]
```diff
@@ -308,6 +308,23 @@ private void checkParseInfixCast(String sqlType) {
     sql(sql).ok(expected);
   }
 
+  @Test public void testCreateTablePermutedColumnLevelAttributes() {
```
Comment (dasch-1) [inline babel/src/test/java/org/apache/calcite/test/BabelParserTest.java:311]: We should add test cases where only a subset of the attributes are explicitly specified (e.g. just "uppercase" or "not uppercase", but none of the others).

File: core/src/main/java/org/apache/calcite/sql/SqlDataTypeSpec.java [hunk 7]
```diff
@@ -83,7 +89,8 @@
   public SqlDataTypeSpec(
       final SqlTypeNameSpec typeNameSpec,
       SqlParserPos pos) {
-    this(typeNameSpec, null, null, pos);
+    this(typeNameSpec, /*timeZone=*/ null, /*nullable=*/ null, /*uppercase=*/ null,
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/SqlDataTypeSpec.java:92]: Thanks for adding these!

File: core/src/main/java/org/apache/calcite/sql/SqlDataTypeSpec.java [hunk 8]
```diff
@@ -97,27 +104,34 @@ public SqlDataTypeSpec(
...
       SqlParserPos pos) {
-    this(typeNameSpec, timeZone, null, pos);
+    this(typeNameSpec, timeZone, /*nullable=*/ null, /*uppercase=*/ null,
+        /*caseSpecific=*/ null, pos);
   }
 
   /**
    * Creates a type specification representing a type, with time zone,
-   * nullability and base type name specified.
+   * nullability, uppercase status, caseSpecific status, and base type name specified.
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/SqlDataTypeSpec.java:113]: nit: wrap this line to 80 characters

File: core/src/main/java/org/apache/calcite/sql/SqlDataTypeSpec.java [hunk 9]
```diff
@@ -97,27 +104,34 @@ public SqlDataTypeSpec(
...
    *
    * @param typeNameSpec The type name can be basic sql type, row type,
    *                     collections type and user defined type
    * @param timeZone     Specified time zone
    * @param nullable     The nullability
+   * @param uppercase    The uppercase status
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/SqlDataTypeSpec.java:119]: The existing param comments are admittedly already pretty weak, but I feel like we can add a little more information here :)

File: core/src/main/java/org/apache/calcite/sql/SqlDataTypeSpec.java (hunk 1 above)
Comment (dasch-1) [reply core/src/main/java/org/apache/calcite/sql/SqlDataTypeSpec.java:199]: Also, the indentation looks a little off.

File: core/src/main/java/org/apache/calcite/sql/SqlDataTypeSpec.java [hunk 10]
```diff
@@ -153,13 +167,42 @@ public Boolean getNullable() {
...
     if (Objects.equals(nullable, this.nullable)) {
       return this;
     }
-    return new SqlDataTypeSpec(typeNameSpec, timeZone, nullable, getParserPosition());
+    return new SqlDataTypeSpec(typeNameSpec, timeZone, nullable, uppercase,
+        caseSpecific, getParserPosition());
+  }
+
+  /** Returns a copy of this data type specification with a given
+   * uppercase status. */
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/SqlDataTypeSpec.java:189]: Indentation?

File: babel/src/test/java/org/apache/calcite/test/BabelParserTest.java (hunk 6 above)
Comment (dasch-1) [reply babel/src/test/java/org/apache/calcite/test/BabelParserTest.java:311]: With these new tests, we can verify that attributes which aren't specified are not written out during unparse.

**Pull request merged**